### PR TITLE
Ghost roles

### PR DESCRIPTION
**Pull request by Ultradragon005**
- [x] remove unneccesary strings.
- [x] Fix evilspirit 
- [X] test, works with players?
- [X] Add to aftermeetingtasks (Added to Exiled Patch)
- [X] Ignore ghostrole in gameresults.

### CODE REVIEW

File: Patches/PlayerControlPatch.cs [hunk 1]
```diff
@@ -1310,6 +1310,14 @@ public static void Postfix(PlayerControl __instance, [HarmonyArgument(0)] Player
             PlagueDoctor.OnAnyMurder();
         }
 
+
+        if (Options.GhostAssign(CustomRoles.GuardianAngelTOHE))
+        {
+            target.RpcSetRole(RoleTypes.GuardianAngel);
+            target.RpcSetCustomRole(CustomRoles.GuardianAngelTOHE);
+        }
+
```
Comment (Ultradragon005) [inline Patches/PlayerControlPatch.cs:1319]: Will add a crewmate_check once I'm done with properly setting it up (Ghost roles will be made in such a way that the first x crewmates to die get them)

File: Modules/OptionHolder.cs [hunk 2]
```diff
@@ -3821,6 +3835,38 @@ public static void SetupRoleOptions(int id, TabGroup tab, CustomRoles role, Cust
...
+            .SetValueFormat(OptionFormat.Players)
+            .SetGameMode(customGameMode);
+
+        if(CustomGhostRoleCounts.ContainsKey(role))
+            CustomRoleCounts.Remove(role);
+
+
+        //CustomRoleSpawnChances.Add(role, spawnOption);
+        CustomRoleCounts.Add(role, countOption);
+        //if (role.RoleExist())
```
Comment (Ultradragon005) [inline Modules/OptionHolder.cs:3854]: So basically find a check for https://i.gyazo.com/cad823b1cb3a003066eaafbc690ce7a6.gif here.

File: Patches/PlayerControlPatch.cs (hunk 1 above)
Comment (Ultradragon005) [reply Patches/PlayerControlPatch.cs:1319]: And also neutrals I think, unless I can change impostor team to make guardian angel impostor compatible.

File: Modules/OptionHolder.cs (hunk 2 above)
Comment (Ultradragon005) [reply Modules/OptionHolder.cs:3854]: I might just make another string option which is "Enabled - Disabled" 😂

**Comment by Ultradragon005** [comment]
image
okay, I know what to do now I think. I will probably make it work tomorrow

**Comment by NikoCat233** [comment]
Patch check protect to accept calls from GuardianAnge

**Comment by Ultradragon005** [comment]
time to find all the instances add-ons prevent spawning <:hehe_sob:1135754965260972123>

**Comment by Ultradragon005** [comment]
okay.... so RoleExist() probably won't work. so I will have make a custom IsEnable method for ghost using Option.GhostAssign()

File: Patches/PlayerControlPatch.cs [hunk 3]
```diff
@@ -1310,6 +1316,40 @@ public static void Postfix(PlayerControl __instance, [HarmonyArgument(0)] Player
...
+                    killer.RpcSetCustomRole(CustomRoles.Warden);
+                    break;
+
+                default:
+                    break;
+
+            }
+
+        }
+
```
Comment (Ultradragon005) [inline Patches/PlayerControlPatch.cs:1347]: If this will be causing issues, I will make a custom harmony checkdeath prefix patch

**Comment by Ultradragon005** [comment]
semi works now (3 hours troubleshooting just to find out I had to use rpcsetrole<:hehe_sob:1135754965260972123> well... good thing I atleast found rpcset role harmony patch bcuz of that to cancel it setting to crewmateghost)

**Comment by NikoCat233** [comment]
bro you can always ask...

File: Modules/OptionHolder.cs [hunk 4]
```diff
@@ -3821,6 +3847,43 @@ public static void SetupRoleOptions(int id, TabGroup tab, CustomRoles role, Cust
...
+        trueghostenable = false;
+        if(CustomGhostRoleCounts.ContainsKey(role)) // Just incase
+            CustomGhostRoleCounts.Remove(role);
+        
+
+        CustomRoleSpawnChances.Add(role, spawnOption);
+        CustomRoleCounts.Add(role, countOption);
+        GetRoleSpawnMode(role);
+        
+        if (trueghostenable)
```
Comment (Ultradragon005) [inline Modules/OptionHolder.cs:3870]: Need help here, this dosen't work for some reason... I need help fidning a way to check if the setting is (0 or 1) before it get sent to all the other places (to prevent ghost-role from spawning)

Comment (Ultradragon005) [reply Modules/OptionHolder.cs:3870]: refer to "GetRoleSpawnMode" method.

Comment (Ultradragon005) [reply Modules/OptionHolder.cs:3870]: nvm, I think Ik the problem, it only runs after game starts I think

Comment (Ultradragon005) [reply Modules/OptionHolder.cs]: okay... so ongamestart patch I go trough a foreach (var plrrole in customrolescounts) which insttantly leaves if not ghost role. And this finally made it possible to detect on/off option.

LMK if you have any better solution (PS: doing almost anything in optionholder fucks shit up. and RandomSpawnPatch is run before Load() so I can't use setupghostroleoptions)

Comment (NikoCat233) [reply Modules/OptionHolder.cs]: Sorry but why do you need a "setupghostroleoptions"

Comment (Ultradragon005) [reply Modules/OptionHolder.cs]: > Sorry but why do you need a "setupghostroleoptions"

To be able and add them to the tab, and disable them from spawning normally, ghost roles will only spawn after a player has died. (Similar to setupadtroles, but I coulnd't quite find out how that disables spawing. So I found another method to do it)

Comment (Ultradragon005) [reply Modules/OptionHolder.cs]: wait... I'm gonna look at how the code looks now, you maybe right and it's uneccessaary, but I think not because I wouldn't use setuadtroleoption, but I'll see...

**Comment by NikoCat233** [comment]
You can just add a isGhostRole in Custom role helper
Use isghostrole=> continue in custom role selector
And assign the ghost role in murder player patch

**Comment by Ultradragon005** [comment]
> You can just add a isGhostRole in Custom role helper Use isghostrole=> continue in custom role selector And assign the ghost role in murder player patch

okay, I will look into it.

**Comment by Ultradragon005** [comment]
OHHH, I didn't know Continue; skips the foreach iteration, alrright the code will be much simpler after I make some changes 😅

**Comment by Ultradragon005** [comment]
about 95% done, just gotta remove unneccesary code I accidentaly added from my previous attempts. And fix evilspirit.

And test with players (& also see if modded clients needs extra rpc's)

**Comment by Ultradragon005** [comment]
@Tommy-XL all works! Also I found out impostor/neutral ghost roles are also possible, but first let's test how Warden works, incase there's any bugs, but so far in my test, there was none.

**Comment by Ultradragon005** [comment]
I will add setupsingleoptionroles in a pr after this has already been merged.

**Comment by Ultradragon005** [comment]
Summary:
- Added ghost roles functionaly.
- Added Crewmate [Ghost] role Warden.
- Fixed Spiritcaller (haven't moved it out of experimental yet, bcuz the code of it was never tested in the first place 💀)

**Comment by Tommy-XL** [comment]
I think that merge is after the release of 1.5.0
Because we need to focus on fixing bugs rather than adding something new

**Comment by Tommy-XL** [comment]
Fix conflicts pls

**Comment by Tommy-XL** [comment]
Can you create a Ghost folder and create two more Impostor/Crewmate folders there?

image

**Comment by Ultradragon005** [comment]
@Tommy-XL done!

**Comment by Tommy-XL** [comment]
Like here

image

**Comment by Ultradragon005** [comment]
> Like here
> 
> image

oh wait I knwo what you mean now.... Ig I will test that, I haven't tested that specific point, since the thing for evilspirit didn't seem for me to work anyways... so ignored it... but I will test

**Comment by Ultradragon005** [comment]
@Tommy-XL I change rpcsetrole patch to be the old one, altough I did not find any issues, hence it was there, I basically replace .is(evilspirit) with ALL ghost roles. So now it is fine.

**Comment by Ultradragon005** [comment]
so, is that all?
